### PR TITLE
Add support for ExposedPorts to the Image config

### PR DESCRIPTION
**Pull request by baronfel**
This is a fairly mechanical transformation.  Things remaining to do:

* tests to verify the addition of the default aspnetcore port declarations (or not, as appropriate)
* tests to verify deduping
* add warnings for malformed ContainerPort Items

### CODE REVIEW

**Comment by baronfel** [comment]
Part of #93 and #112

**Comment by baronfel** [comment]
For a test asp.net core container, we now have ExposedPorts:

```json
"ExposedPorts": {
    "80/tcp": {}
}
```

and we can see the ports being automatically bound when `-P` is used:

```shell
➜ docker ps
CONTAINER ID   IMAGE                      COMMAND                  CREATED         STATUS         PORTS                   NAMES
a41168f7eac0   container-app-tres:0.0.4   "/app/container-app-…"   5 seconds ago   Up 4 seconds   0.0.0.0:49153->80/tcp   wonderful_faraday
```

and indeed, navigating to that port shows the app.

File: Microsoft.NET.Build.Containers/ContainerHelpers.cs [hunk 1]
```diff
@@ -136,11 +136,27 @@ public static bool NormalizeImageName(string containerImageName, [NotNullWhen(fa
...
         }
     }
+
+    public static bool TryParsePort(string input, [NotNullWhen(true)] out Port? port)
+    {
+        var parts = input.Split('/');
+        Port? p =
+            parts switch
+            {
+                [var portNumber, var type] when int.TryParse(portNumber, out var portInt) && Enum.TryParse<PortType>(type, out var portType) => new(portInt, portType),
```
Comment (baronfel) [inline Microsoft.NET.Build.Containers/ContainerHelpers.cs:155]: list/array patterns are lovely, I'm so glad they're in C# now.

File: Microsoft.NET.Build.Containers/ContainerHelpers.cs [hunk 2]
```diff
@@ -136,11 +136,28 @@ public static bool NormalizeImageName(string containerImageName, [NotNullWhen(fa
...
+            else if (!Char.IsLetterOrDigit(containerImageName, 0))
+            {
                 throw new ArgumentException("The first character of the image name must be a lowercase letter or a digit.");
             }
             normalizedImageName = imageNameCharacters.Replace(containerImageName, "-");
             return false;
         }
     }
+
+    public static bool TryParsePort(string input, [NotNullWhen(true)] out Port? port)
```
Comment (BenVillalobos) [inline Microsoft.NET.Build.Containers/ContainerHelpers.cs:149]: Can we get a test for this?

File: Microsoft.NET.Build.Containers/ContainerHelpers.cs [hunk 3]
```diff
@@ -136,11 +136,28 @@ public static bool NormalizeImageName(string containerImageName, [NotNullWhen(fa
...
+    {
+        var parts = input.Split('/');
+        Port? p =
+            parts switch
+            {
+                [var portNumber, var type] when int.TryParse(portNumber, out var portInt)
```
Comment (BenVillalobos) [inline Microsoft.NET.Build.Containers/ContainerHelpers.cs:155]: I appreciate you using the newer c# features!

Thinking out loud as I make sense of it:

You switch over `parts` which is a list of items split on `/`.

If `parts` is two elements long, parse them and return a new Port if you could parse.
If you got a single element, parse the port and default to tcp.
Whatever else you got, abandon all hope.

The first case is hard to read if it's you first time looking it over, but no big deal once you learn it. I like it.

File: Microsoft.NET.Build.Containers/build/Microsoft.NET.Build.Containers.targets [hunk 4]
```diff
@@ -76,6 +76,7 @@
                         WorkingDirectory="$(ContainerWorkingDirectory)"
                         Entrypoint="@(ContainerEntrypoint)"
                         EntrypointArgs="@(ContainerEntrypointArgs)"
-						Labels="@(ContainerLabel)"/>
+                        Labels="@(ContainerLabel)"
+                        ExposedPorts="@(ContainerPort)" />
```
Comment (BenVillalobos) [inline Microsoft.NET.Build.Containers/build/Microsoft.NET.Build.Containers.targets:80]: Nit: We may want to rename `ContainerPort` to `ContainerPorts`.

File: Microsoft.NET.Build.Containers/CreateNewImage.cs [hunk 5]
```diff
@@ -122,7 +130,17 @@ public override bool Execute()
         {
             image.Label(label.ItemSpec, label.GetMetadata("Value"));
         }
-        
+
+        foreach (var port in ExposedPorts)
+        {
+            if (int.TryParse(port.ItemSpec, out int portNumber)
```
Comment (BenVillalobos) [inline Microsoft.NET.Build.Containers/CreateNewImage.cs:136]: Shouldn't tryparseport happen here? We can log a warning or something if it didn't succeed, and `Image` can assume what it was given would work (or we do the validation both places). I feel like customers should get some explicit message from us if something goes wrong here though. Thoughts?

**Review by BenVillalobos — APPROVED** [review]
overall lgtm, my biggest gripe is testing. Won't block on it if we need to merge & create an issue for that

File: Microsoft.NET.Build.Containers/ContainerHelpers.cs (hunk 3 above)
Comment (baronfel) [reply Microsoft.NET.Build.Containers/ContainerHelpers.cs:155]: This actually looks really nice in F# because patterns compose in a way that they currently don't in C#:

```fsharp
match p.Split("/") with
| [ Int portNum; Enum<PortType> portTye ] -> ...
| [Int portNum ] -> ...
| _ -> ...
```
you can nest patterns inside other patterns easily, including user-defined patterns (since a pattern is just a function).

File: Microsoft.NET.Build.Containers/build/Microsoft.NET.Build.Containers.targets (hunk 4 above)
Comment (baronfel) [reply Microsoft.NET.Build.Containers/build/Microsoft.NET.Build.Containers.targets:80]: excellent point 👍

File: Microsoft.NET.Build.Containers/CreateNewImage.cs (hunk 5 above)
Comment (baronfel) [reply Microsoft.NET.Build.Containers/CreateNewImage.cs:136]: I agree on all counts, but the problem for TryParsePort is that they want the whole string. I suppose we could concat the ItemSpec and Type, if it exists?

File: Microsoft.NET.Build.Containers/ContainerHelpers.cs (hunk 2 above)
Comment (baronfel) [reply Microsoft.NET.Build.Containers/ContainerHelpers.cs:149]: can do!

File: Microsoft.NET.Build.Containers/build/Microsoft.NET.Build.Containers.targets (hunk 4 above)
Comment (baronfel) [reply Microsoft.NET.Build.Containers/build/Microsoft.NET.Build.Containers.targets:80]: Thinking about this more I don't think I agree - each individual MSBuild Item is a single `ContainerPort`.

File: Microsoft.NET.Build.Containers/ContainerHelpers.cs (hunk 2 above)
Comment (baronfel) [reply Microsoft.NET.Build.Containers/ContainerHelpers.cs:149]: Excellent reminder to test because I'd inverted the boolean condition at the end of the method.

File: Microsoft.NET.Build.Containers/CreateNewImage.cs (hunk 5 above)
Comment (baronfel) [reply Microsoft.NET.Build.Containers/CreateNewImage.cs:136]: Another problem with TryParsePort - it's parsing what should be a machine-written value in the container config. This is validating user data, so we are probably ok to not try to reuse that too much and instead be more specific about the user errors here.

Comment (baronfel) [reply Microsoft.NET.Build.Containers/CreateNewImage.cs:136]: I, uh, _may_ have done some refactoring here so that both port-parsing points use a shared parsing logic.

**Comment by baronfel** [comment]
I made significant enough changes to the port parsing to warrant a re-review, but I did add a crap ton of testing to the core logic. Also went ham on the user-facing errors.

File: Microsoft.NET.Build.Containers/Image.cs [hunk 6]
```diff
@@ -7,6 +8,14 @@ namespace Microsoft.NET.Build.Containers;
 
 record Label(string name, string value);
 
+public enum PortType
+{
+    tcp,
+    udp
```
Comment (rainersigwald) [inline Microsoft.NET.Build.Containers/Image.cs:14]: .NET style guide says these should start with a capital letter, but I do not feel super strongly about it.

File: Microsoft.NET.Build.Containers/ContainerHelpers.cs [hunk 7]
```diff
@@ -141,4 +141,64 @@ public static bool NormalizeImageName(string containerImageName, [NotNullWhen(fa
...
+    [Flags]
+    public enum ParsePortError
+    {
+        MissingPortNumber,
+        InvalidPortNumber,
+        InvalidPortType,
+        UnknownPortFormat
+    }
+
+    public record ParsePortResult(bool success, Port? port, ParsePortError? parseErrors);
```
Comment (rainersigwald) [inline Microsoft.NET.Build.Containers/ContainerHelpers.cs:154]: This was hard to follow for me; in C# I think I'd prefer it outside the type system:

```c#
bool TryParsePort(string input, [NotNullWhen(true)] Port? port, [NotNullWhen(false)] ParsePortError? error)
```

File: Microsoft.NET.Build.Containers/Image.cs [hunk 8]
```diff
@@ -84,11 +110,35 @@ private static HashSet<Label> ReadLabelsFromConfig(JsonNode inputConfig)
...
             return new HashSet<Label>();
         }
     }
 
+    private static HashSet<Port> ReadPortsFromConfig(JsonNode inputConfig)
+    {
+        if (inputConfig is JsonObject config && config["ExposedPorts"] is JsonObject portsJson)
+        {
+            // read label mappings from object
+            var ports = new HashSet<Port>();
```
Comment (rainersigwald) [inline Microsoft.NET.Build.Containers/Image.cs:123]: Think it'll be a problem that HashSet isn't order-preserving? It doesn't seem like it would be to me . . .

File: Microsoft.NET.Build.Containers/ContainerHelpers.cs [hunk 9]
```diff
@@ -141,4 +141,64 @@ public static bool NormalizeImageName(string containerImageName, [NotNullWhen(fa
...
+        if (String.IsNullOrEmpty(portNumber))
+        {
+            errors = ParsePortError.MissingPortNumber;
+        }
+        else if (!int.TryParse(portNumber, out port))
+        {
+            errors = ParsePortError.InvalidPortNumber;
+        }
+
+        if (!Enum.TryParse<PortType>(portType, out PortType t))
```
Comment (rainersigwald) [inline Microsoft.NET.Build.Containers/ContainerHelpers.cs:169]: Is the spec case sensitive?

```suggestion
        if (!Enum.TryParse(portType, ignoreCase: true, out PortType t))
```
?

File: Microsoft.NET.Build.Containers/CreateNewImage.cs [hunk 10]
```diff
@@ -154,7 +219,14 @@ public override bool Execute()
 
         if (BuildEngine != null)
         {
-            Log.LogMessage(MessageImportance.High, "Pushed container '{0}:{1}' to registry '{2}'", ImageName, ImageTag, OutputRegistry);
+            if (IsDockerPush)
```
Comment (rainersigwald) [inline Microsoft.NET.Build.Containers/CreateNewImage.cs:222]: Nit: this should be in a different commit/PR.

File: Microsoft.NET.Build.Containers/CreateNewImage.cs [hunk 11]
```diff
@@ -84,8 +93,56 @@ public CreateNewImage()
...
+            var portNo = port.ItemSpec;
+            var portTy = port.GetMetadata("Type");
+            var parsePortResult = ContainerHelpers.ParsePort(portNo, portTy);
+            if (!parsePortResult.success)
+            {
+                ContainerHelpers.ParsePortError errors = (ContainerHelpers.ParsePortError)parsePortResult.parseErrors!;
+                var portString = portTy == null ? portNo : $"{portNo}/{portTy}";
+                if (errors.HasFlag(ContainerHelpers.ParsePortError.MissingPortNumber))
+                {
+                    Log.LogError("A ContainerPort item was provided without an Include metadata specifying the port number. Please provide a ContainerPort item with an ItemSpec: <ContainerPort Include=\"80\" />");
```
Comment (rainersigwald) [inline Microsoft.NET.Build.Containers/CreateNewImage.cs:112]: Important for diagnosability IMO to provide the bogus input in the error:

```suggestion
                    Log.LogError("ContainerPort item '{0}' does not specify the port number. Please ensure the item's Include is a port number, for example '<ContainerPort Include=\"80\" />'", port.ItemSpec);
```

File: Microsoft.NET.Build.Containers/Image.cs (hunk 6 above)
Comment (baronfel) [reply Microsoft.NET.Build.Containers/Image.cs:14]: Definitely doable, I'd just worry about the parsing side - the spec is case-sensitive for these values.

File: Test.Microsoft.NET.Build.Containers/ContainerHelpersTests.cs [hunk 12]
```diff
@@ -75,4 +75,27 @@ public void IsValidImageTag_InvalidLength()
...
+    [DataRow("/tcp", false, null, null, ContainerHelpers.ParsePortError.MissingPortNumber)]
+    public void CanParsePort(string input, bool shouldParse, int? expectedPortNumber, PortType? expectedType, ContainerHelpers.ParsePortError? expectedError) {
+        var parseSuccess = ContainerHelpers.TryParsePort(input);
+        Assert.AreEqual<bool>(shouldParse, parseSuccess.success, $"{(shouldParse ? "Should" : "Shouldn't")} have parsed {input} into a port");
+        if (!shouldParse) {
+            Assert.IsNull(parseSuccess.port);
+            Assert.IsNotNull(parseSuccess.parseErrors);
+            Assert.AreEqual(expectedError, parseSuccess.parseErrors);
+        }
+        if (shouldParse) {
```
[the remark this answers is not in the document]
Comment (baronfel) [reply Test.Microsoft.NET.Build.Containers/ContainerHelpersTests.cs:91]: yep, done

File: Microsoft.NET.Build.Containers/Image.cs (hunk 6 above)
Comment (baronfel) [reply Microsoft.NET.Build.Containers/Image.cs:16]: added comments as to _why_ they are lowercased.

File: Microsoft.NET.Build.Containers/ContainerHelpers.cs (hunk 7 above)
Comment (baronfel) [reply Microsoft.NET.Build.Containers/ContainerHelpers.cs]: refactored to this.

File: Microsoft.NET.Build.Containers/ContainerHelpers.cs (hunk 9 above)
Comment (baronfel) [reply Microsoft.NET.Build.Containers/ContainerHelpers.cs:168]: it is, I've noted so on the enum

**Review by rainersigwald — APPROVED** [review]

File: Microsoft.NET.Build.Containers/Image.cs (hunk 8 above)
Comment (baronfel) [reply Microsoft.NET.Build.Containers/Image.cs:123]: Agree - the underlying collection (a JS object) isn't ordered either

**Comment by baronfel** [comment]
@BenVillalobos when this merges we'll want to take care and make sure that the new tool task also handles writing port arguments to the CLI.
